function in struct field

The following program `sample.go` triggers a panic:

```go
package main

type SecretProvider func(user, realm string) string

type BasicAuth struct {
    Realm   string
    Secrets SecretProvider
}

func (a *BasicAuth) CheckAuth() string { return a.Secrets("me", a.Realm) }

func (a *BasicAuth) secretBasic(user, realm string) string { return a.Realm + "-" + user + "-" + realm }

func main() {
    b := &BasicAuth{Realm: "test"}
    b.Secrets = b.secretBasic
    s := b.CheckAuth()
    println(s)
}
```

Expected result:
```console
$ go run ./sample.go
test-me-test
```

Got:
```console
$ yaegi ./sample.go
panic: reflect.Set: value of type *interp.Node is not assignable to type func(string, string) string [recovered]
	panic: reflect.Set: value of type *interp.Node is not assignable to type func(string, string) string

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc0001725f0)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:106 +0x134
panic(0x17fb640, 0xc000158f10)
	/usr/local/Cellar/go/1.12.6/libexec/src/runtime/panic.go:522 +0x1b5
reflect.Value.assignTo(0x188e1c0, 0xc0001ece00, 0x16, 0x191cb22, 0xb, 0x1828380, 0x0, 0x100, 0x1905c00, 0xc0001b1d01)
	/usr/local/Cellar/go/1.12.6/libexec/src/reflect/value.go:2339 +0x437
reflect.Value.Set(0x1828380, 0xc00015d0d0, 0x193, 0x188e1c0, 0xc0001ece00, 0x16)
	/usr/local/Cellar/go/1.12.6/libexec/src/reflect/value.go:1473 +0xa8
github.com/containous/yaegi/interp.assign.func3(0xc0001725f0, 0xc00015d000)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:243 +0xb6
github.com/containous/yaegi/interp.runCfg(0xc0001eba00, 0xc0001725f0)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:111 +0x62
github.com/containous/yaegi/interp.(*Interpreter).run(0xc0000901e0, 0xc0001eaf00, 0xc000172050)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:93 +0x126
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0000901e0, 0xc0001dc000, 0x1cb, 0x3cb, 0xc0001dc000, 0x1cb, 0x0, 0xc0001bc680)
	/Users/marc/go/src/github.com/containous/yaegi/interp/interp.go:283 +0x32d
main.main()
	/Users/marc/go/src/github.com/containous/yaegi/yaegi.go:61 +0x443
```

--- a/_test/struct20.go
+++ b/_test/struct20.go
@@ -1,0 +1,21 @@
+package main
+
+type SecretProvider func(user, realm string) string
+
+type BasicAuth struct {
+	Realm   string
+	Secrets SecretProvider
+}
+
+func (a *BasicAuth) CheckAuth() string { return a.Secrets("me", a.Realm) }
+
+func secretBasic(user, realm string) string { return user + "-" + realm }
+
+func main() {
+	b := &BasicAuth{"test", secretBasic}
+	s := b.CheckAuth()
+	println(s)
+}
+
+// Output:
+// me-test

--- a/_test/struct21.go
+++ b/_test/struct21.go
@@ -1,0 +1,22 @@
+package main
+
+type SecretProvider func(user, realm string) string
+
+type BasicAuth struct {
+	Realm   string
+	Secrets SecretProvider
+}
+
+func (a *BasicAuth) CheckAuth() string { return a.Secrets("me", a.Realm) }
+
+func (a *BasicAuth) secretBasic(user, realm string) string { return a.Realm + "-" + user + "-" + realm }
+
+func main() {
+	b := &BasicAuth{Realm: "test"}
+	b.Secrets = b.secretBasic
+	s := b.CheckAuth()
+	println(s)
+}
+
+// Output:
+// test-me-test

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1151,6 +1151,11 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 				case PtrT:
 					n.typ = n.typ.fieldSeq(ti)
 					n.gen = getPtrIndexSeq
+					if n.typ.cat == FuncT {
+						// function in a struct field is always wrapped in reflect.Value
+						rtype := n.typ.TypeOf()
+						n.typ = &Type{cat: ValueT, rtype: rtype}
+					}
 				default:
 					n.gen = getIndexSeq
 					n.typ = n.typ.fieldSeq(ti)

--- a/interp/run.go
+++ b/interp/run.go
@@ -8,6 +8,8 @@ import (
 	"reflect"
 )
 
+func init() { log.SetFlags(log.Lshortfile) }
+
 // Builtin type defines functions which run at CFG execution
 type Builtin func(f *Frame) Builtin
 
@@ -400,7 +402,11 @@ func _panic(n *Node) {
 }
 
 func genFunctionWrapper(n *Node) func(*Frame) reflect.Value {
-	def := n.val.(*Node)
+	var def *Node
+	var ok bool
+	if def, ok = n.val.(*Node); !ok {
+		return genValueAsFunctionWrapper(n)
+	}
 	setExec(def.child[3].start)
 	start := def.child[3].start
 	numRet := len(def.typ.ret)
@@ -1446,12 +1452,13 @@ func compositeSparse(n *Node) {
 	values := make(map[int]func(*Frame) reflect.Value)
 	a, _ := n.typ.zero()
 	for _, c := range child {
+		c1 := c.child[1]
 		field := n.typ.fieldIndex(c.child[0].ident)
-		convertLiteralValue(c.child[1], n.typ.field[field].typ.TypeOf())
-		if c.typ.cat == FuncT {
-			values[field] = genFunctionWrapper(c.child[1])
+		convertLiteralValue(c1, n.typ.field[field].typ.TypeOf())
+		if c1.typ.cat == FuncT {
+			values[field] = genFunctionWrapper(c1)
 		} else {
-			values[field] = genValue(c.child[1])
+			values[field] = genValue(c1)
 		}
 	}
 

--- a/interp/value.go
+++ b/interp/value.go
@@ -44,6 +44,13 @@ func genValueRecv(n *Node) func(*Frame) reflect.Value {
 	}
 }
 
+func genValueAsFunctionWrapper(n *Node) func(*Frame) reflect.Value {
+	v := genValue(n)
+	return func(f *Frame) reflect.Value {
+		return genFunctionWrapper(v(f).Interface().(*Node))(f)
+	}
+}
+
 func genValueAs(n *Node, t reflect.Type) func(*Frame) reflect.Value {
 	v := genValue(n)
 	return func(f *Frame) reflect.Value {

